File: 378. Kth Smallest Element in a Sorted Matrix.py
class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        '''
        Using heap to pick first n smallest items
        Time O(klogn)
        Space O(n)
        '''
        heap = []
        n = len(matrix)
        for i in range(n):
            heapq.heappush(heap, (matrix[0][i], 0, i))
        for i in range(k):
            num, row, col = heapq.heappop(heap)
            if row < n-1:
                heapq.heappush(heap, (matrix[row+1][col], row+1, col))
            
        return num

class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        '''
        Binary Search
        Time O(N∗log(max−min))
        Space O(1)
        '''
        n = len(matrix)
        start, end = matrix[0][0], matrix[n - 1][n - 1]
        while start < end:
            mid = start + (end - start) // 2
            smaller, larger = (matrix[0][0], matrix[n - 1][n - 1])

            count, smaller, larger = self.count_less_equal(matrix, mid, smaller, larger)

            if count == k:
                return smaller
            if count < k:
                start = larger  # search higher
            else:
                end = smaller  # search lower

        return start


    def count_less_equal(self, matrix, mid, smaller, larger):
        count, n = 0, len(matrix)
        row, col = n - 1, 0
        while row >= 0 and col < n:
            if matrix[row][col] > mid:
        # as matrix[row][col] is bigger than the mid, let's keep track of the
        # smallest number greater than the mid
                larger = min(larger, matrix[row][col])
                row -= 1
            else:
        # as matrix[row][col] is less than or equal to the mid, let's keep track of the
        # biggest number less than or equal to the mid
                smaller = max(smaller, matrix[row][col])
                count += row + 1
                col += 1

        return count, smaller, larger

File: 378. Kth Smallest Element in a Sorted Matrix.java
class Solution {
    /**
     * This use heap to poll the smallest from row and then add the next column into the heap,
     * this could work since each row and each column is sorted.
     * Time klog(m)
     * @param matrix
     * @param k
     * @return
     */
    public int kthSmallest(int[][] matrix, int k) {
        if (matrix==null || matrix.length==0 || matrix[0].length==0) {
            return 0;
        }
        int row = matrix.length, col = matrix[0].length;
        PriorityQueue<Tuple> pq = new PriorityQueue<>();
        for (int i=0; i<row; i++) {
            pq.offer(new Tuple(matrix[i][0], i, 0));
        }
        while (k > 1) {
            Tuple temp = pq.poll();
            if (temp.col < col-1) {
                pq.offer(new Tuple(matrix[temp.row][temp.col+1], temp.row, temp.col+1));
            }
            k--;
        }
        return pq.poll().val;
    }
}

class Tuple implements Comparable<Tuple> {
    int val;
    int row;
    int col;
    
    public Tuple(int val, int row, int col) {
        this.val = val;
        this.row = row;
        this.col = col;
    }
    
    @Override
    public int compareTo(Tuple that) {
        return this.val - that.val;
    }
}

/**
 * This is a binary search version,
 * key point is to figure out the search region
 * in this question, the search region is not the index, but the range from the smallest to the largest
 * and in each part we calculate the count of the number smaller than mid
 */
class Solution {
    public int kthSmallest(int[][] matrix, int k) {
        if (matrix==null || matrix.length==0 || matrix[0].length==0) {
            return 0;
        }
        int row = matrix.length, col = matrix[0].length;
        int lo = matrix[0][0], hi = matrix[row-1][col-1]+1;
        while (lo < hi) {
            int mid = (hi-lo)/2+lo;
            int count = 0;
            int j = col-1;
            for (int i=0; i<row; i++) {
                while (j>=0 && matrix[i][j] > mid) {j--;}
                count += j+1;
            }
            if (count < k) {
                lo = mid+1;
            } else {
                hi = mid;
            }
        }
        return lo;
    }
}